### PR TITLE
Revert "Revert D30752939: [pytorch][PR] nvfuser update"

### PR DESCRIPTION
**Pull request by jjsjann123**
This reverts commit 03389dc851db6f3ca52f9a4455ce2090c64a223d.

Attempt again for PR: #63745
Fixes the windows build failure.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 1]
```diff
@@ -32,7 +32,7 @@ Fusion* FusionGuard::getCurFusion() {
   return ACTIVE_FUSION;
 }
 
-void swap(Fusion& a, Fusion& b) noexcept {
+TORCH_CUDA_CU_API void swap(Fusion& a, Fusion& b) noexcept {
```
Comment (malfet) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:35]: Windows compilation fails on this line with:
```
2021-09-16T10:43:07.5518815Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch\csrc\jit\codegen\cuda\fusion.cpp(35): error C2375: 'torch::jit::fuser::cuda::swap': redefinition; different linkage
2021-09-16T10:43:07.5521053Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch/csrc/jit/codegen/cuda/ir_base_nodes.h(63): note: see declaration of 'torch::jit::fuser::cuda::swap'
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:35]: Thanks! missed that one in the log :pleading_face:

**Comment by malfet** [comment]
Here is the list of compilation failures on Windows:
```
2021-09-16T10:43:02.9236331Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\xutility(4393): error C2280: 'std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>> &std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>>::operator =(const std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>> &)': attempting to reference a deleted function
2021-09-16T10:43:07.5518815Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch\csrc\jit\codegen\cuda\fusion.cpp(35): error C2375: 'torch::jit::fuser::cuda::swap': redefinition; different linkage
2021-09-16T10:43:07.5523016Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch\csrc\jit\codegen\cuda\fusion.cpp(120): error C2065: 'swap': undeclared identifier
2021-09-16T10:43:07.5524974Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch\csrc\jit\codegen\cuda\fusion.cpp(127): error C2065: 'swap': undeclared identifier
2021-09-16T10:43:07.5526942Z C:\actions-runner\_work\pytorch\pytorch\pytorch-1240968219\torch\csrc\jit\codegen\cuda\fusion.cpp(134): error C2065: 'swap': undeclared identifier
2021-09-16T10:43:09.6849038Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\xutility(4393): error C2280: 'std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>> &std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>>::operator =(const std::unique_ptr<torch::jit::fuser::cuda::SchedulerEntry,std::default_delete<torch::jit::fuser::cuda::SchedulerEntry>> &)': attempting to reference a deleted function
```

**Review by seemethere — APPROVED** [review]
LGTM on green

**Comment by wconstab** [comment]
fingers crossed, @jjsjann123!

**Comment by jjsjann123** [comment]
Windows build passed! awesome! Seems like there's something with c10::optional template on non-copyable classes on Windows.

I'll clean up linter issues.

**Comment by jjsjann123** [comment]
FYI just to verify that `linux-bionic-py3.6-clang9 ` seems to be irrelevant and not introduced by this PR, I should ignore that?

**Comment by jjsjann123** [comment]
I'm waiting on CI tests to finish just to boost my confidence.
I have turned off a few python tests for our fuser and will resolve the conflicts from upstream master in the next push.

**Comment by jjsjann123** [comment]
> Here is link to goltbot that narrows it down (i.e. its a deficiency of older VC++ compiler: https://godbolt.org/z/6Prd77xYs )

narrowing conversion can be work around by `static_cast<size_t>()`. That line somehow is missed by clang-tidy.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+    auto welford_out = Welford(x, x_reduction_axes);
+
+    // updating running mean and running var
+    if (running_mean != nullptr && running_var != nullptr) {
+      auto rev_momentum = sub(new Double(1.0), momentum);
+      auto current_mean_hat = mul(welford_out.avg, momentum);
+      auto mean_hat = mul(running_mean, rev_momentum);
+      auto new_mean_hat = add(mean_hat, current_mean_hat);
+
+      auto new_mean_sum = sum(new_mean_hat, {kBatchDim});
```
Comment (malfet) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:500]: Replaced with
```suggestion
      // NS: static_cast to workaround VC++ error, see https://godbolt.org/z/6Prd77xYs
      auto new_mean_sum = sum(new_mean_hat, {static_cast<int>(kBatchDim)});
```

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+      fusion->addOutput(new_mean_channels_only);
+      fusion->aliasOutputToInput(new_mean_channels_only, running_mean);
+
+      auto num_feature_decrement = sub(N, new Int(1));
+      auto unbiased_var = div(welford_out.var_sum, num_feature_decrement);
+      auto current_var_hat = mul(unbiased_var, momentum);
+      auto var_hat = mul(running_var, rev_momentum);
+      auto new_var_hat = add(var_hat, current_var_hat);
+
+      auto new_var_sum = sum(new_var_hat, {kBatchDim});
```
Comment (malfet) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:511]: Replaced with
```suggestion
      // NS: static_cast to workaround VC++ error, see https://godbolt.org/z/6Prd77xYs
      auto new_var_sum = sum(new_var_hat, {static_cast<int>(kBatchDim)});
```

**Comment by malfet** [comment]
@jjsjann123 `clang-tidy` didn't spot it, because this narrowing warning is false positive, as `kBatchDim` is known at compile time, and newer versions of VC++ do not raise this error, as you can check by chaning compiler version in provided godbolt link)

**Review by malfet — APPROVED** [review]

**Comment by jjsjann123** [comment]
> @jjsjann123 `clang-tidy` didn't spot it, because this narrowing warning is false positive, as `kBatchDim` is known at compile time, and newer versions of VC++ do not raise this error, as you can check by changing compiler version in provided godbolt link)

Got ya. Thanks for the fixes.

**Comment by malfet** [comment]
Will land other linter suggested fixes in https://github.com/pytorch/pytorch/pull/65459

**Pull request closed without merging**